File: platforms/apple/Sources/BotaAppleSDK/Bluetooth/EncryptedUploadV2CapabilityReader.swift
import CryptoKit
import Foundation

struct EncryptedUploadV2CapabilityReader: Sendable {
    typealias Read = @Sendable (String, String, String) async throws -> Data
    typealias Decode = @Sendable (Data) throws -> EncryptedUploadV2CapabilitiesValue

    private let read: Read
    private let decode: Decode

    init(read: @escaping Read, decode: @escaping Decode) {
        self.read = read
        self.decode = decode
    }

    func readFresh(peripheralID: String) async throws -> EncryptedUploadV2CapabilitySnapshot {
        let value = try await read(
            peripheralID,
            BotaBluetoothUUIDs.storageService,
            BotaBluetoothUUIDs.storageTransferCapabilitiesV2
        )
        return EncryptedUploadV2CapabilitySnapshot(
            rawValue: value,
            sha256: Data(SHA256.hash(data: value)),
            capabilities: .init(value: try decode(value))
        )
    }
}

File: platforms/android/sdk/src/main/kotlin/dev/bota/sdk/internal/core/EncryptedUploadV2CapabilityReader.kt
package dev.bota.sdk.internal.core

import dev.bota.sdk.EncryptedUploadV2Capabilities
import dev.bota.sdk.EncryptedUploadV2CapabilitySnapshot
import dev.bota.sdk.internal.bluetooth.BotaBluetoothUUIDs
import java.security.MessageDigest
import java.util.UUID

internal class EncryptedUploadV2CapabilityReader(
    private val read: suspend (String, UUID, UUID) -> ByteArray,
    private val decode: (ByteArray) -> EncryptedUploadV2CapabilitiesValue,
) {
    suspend fun readFresh(peripheralId: String): EncryptedUploadV2CapabilitySnapshot {
        val raw = read(
            peripheralId,
            BotaBluetoothUUIDs.StorageService,
            BotaBluetoothUUIDs.TransferCapabilitiesV2,
        )
        val value = decode(raw)
        return EncryptedUploadV2CapabilitySnapshot(
            raw,
            MessageDigest.getInstance("SHA-256").digest(raw),
            EncryptedUploadV2Capabilities(
                value.flags,
                value.maximumSignedBlobBytes,
                value.maximumManifestBytes,
                value.maximumDataPayloadBytes,
                value.maximumWindowPackets,
                value.durableCheckpointIntervalBlocks,
                value.maximumMissingSequences,
            ),
        )
    }
}
